### DISCUSSION TITLE
how to mutate share state in route handlers

### DISCUSSION THREAD

**Discussion by ali322** [tokio-rs/axum#629] [General]
is there some way to mutate share state in route handlers, i had a golang project use package variables to store some share state and mutated it in route handlers, when i try to rewrite this golang project with axum, i have no idea about how to mutate share state, any suggestions?

**Comment by ttys3** [comment]
please see the example https://github.com/tokio-rs/axum/blob/dea36db400f27c025b646e5720b9a6784ea4db6e/examples/key-value-store/src/main.rs

this should help

**Comment by ttys3** [comment]
there's also a real world example in my `static-server` project, which is very simple.

see https://github.com/ttys3/static-server/blob/main/src/main.rs#L162

```rust

	  struct StaticServerConfig {
	      pub(crate) root_dir: String,
	  }

    let app = Router::new()
        .route("/favicon.ico", get(favicon))
        .route("/healthz", get(health_check))
        .fallback(get(index_or_content))
        .layer(AddExtensionLayer::new(Arc::new(StaticServerConfig { root_dir })));
```

and in the handler:

```rust
async fn index_or_content(Extension(cfg): Extension<Arc<StaticServerConfig>>, req: Request<Body>) -> impl IntoResponse {
// ...
}
```

**Comment by davidpdrsn** [comment]
To mutate shared state you’ll need an Arc<Mutex<YourState>>, or equivalent.

**Comment by anna-hope** [comment]
I'm trying to pass mutable shared state to my route handler function using an `Arc<RwLock>` or `Arc<Mutex>`, and I'm getting compiler errors with both.

My code looks like this:

```
use std::sync::{Arc, RwLock};

type SharedState = Arc<RwLock<ConfigManager>>;

async fn configs(Extension(state): Extension<Arc<RwLock<ConfigManager>>>) -> String {
   // mutate state here
   // ...
}

let config_manager = SharedState::default();

let app = Router::new()
    .route("/configs", get(model_configs))
    .layer(Extension(config_manager));
```

This is basically the same as in the example linked by @ttys3, but I'm getting the following compiler error:

```
the trait `Handler<_, _>` is not implemented for `fn(Extension<Arc<std::sync::RwLock<ConfigManager>>>) -> impl Future<Output = String> {configs}

note: required by a bound in `axum::routing::get`

top_level_handler_fn!(get, GET);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `axum::routing::get`
```

Does anyone have any suggestions?

**Reply by davidpdrsn** [reply]
See https://docs.rs/axum/latest/axum/handler/index.html#debugging-handler-type-errors

**Reply by anna-hope** [reply]
@davidpdrsn Thank you! That eventually helped me figure out that the problem was that I was using an `await` on the Mutex'd value. I solved it by using the Mutex from [async-lock](https://crates.io/crates/async-lock).

**Reply by davidpdrsn** [reply]
Great! Tokio also has async locks built-in https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html btw

**Reply by anna-hope** [reply]
Oh, thanks for pointing that out! The one from tokio is definitely preferable since it doesn't involve pulling an extra dependency.
